### DISCUSSION TITLE
Typesafe mdast tree

### DISCUSSION THREAD

**Discussion by Ethan-Arrowood** [remarkjs/remark#649] [Q&A] [answered]
Hello,

I'm trying to build a utility that transforms an mdast tree into my own AST. I have this code so far:

```ts
remark()
  .use(() => {
    return (tree) => {
      console.log(tree)
    }
  })
  .process(
    /* markdown */,
    () => {}
  )
```

How do I get the `tree` to be an `mdast` type? Its currently returning `unist.Node`.

**Comment by Ethan-Arrowood** [comment]
I've added a type guard for now using 'mdast-util-assert':

```ts
function isMdastRoot(tree: unist.Node): tree is mdast.Root {
  return mdastAssert(tree) && tree.type === 'root'
}

remark()
  .use(() => {
    return tree => {
      if (isMdastRoot(tree)) {
        console.log(tree)

      } else {
        throw new Error('Invalid tree - is not mdast root')
      }
    }
  })
```

**Comment by ChristianMurphy** [comment, marked as answer]
A common way is to use https://github.com/syntax-tree/unist-util-is

```ts
import is from 'unist-util-is'
import type { Root } from 'mdast'

// get node from `unified` and/or `remark`

if (is<Root>(node, 'root') {
  // do something with root node
}

**Reply by ChristianMurphy** [reply]
another is to use https://github.com/syntax-tree/unist-util-visit if walking the AST is needed
https://github.com/syntax-tree/unist-util-visit/blob/1e7db9e733c199487bedd2f06033e86327604069/types/unist-util-visit-tests.ts#L54-L67

**Reply by ChristianMurphy** [reply]
In both cases these provide an easy way to select specific node types, and also give type consistency guarantees through some moderately complex generics used under the hood https://github.com/syntax-tree/unist-util-is/blob/bd971234cc4639a49371beefde66697f34902818/index.d.ts#L11-L40

**Reply by Ethan-Arrowood** [reply]
This is what I was looking for thank you!
